File: 2019/py/day_02.py
import itertools as it
from pathlib import Path

from intcode import CPU


def main():
    # Read in the program
    text = (Path(__file__).parent / "../input/input_02.txt").read_text()
    program = [int(x) for x in text.split(',')]
    # Part 1
    cpu = CPU(program.copy())
    cpu.mem[1:3] = 12, 2
    cpu.run()
    print("Part 1:", cpu.mem[0])
    # Part 2: Find the inputs that will produce 19690720
    for noun, verb in it.product(range(100), repeat=2):
        cpu = CPU(program.copy())
        cpu.mem[1:3] = noun, verb
        cpu.run()
        if cpu.mem[0] == 19690720:
            print("Part 2:", 100 * noun + verb)
            return


if __name__ == "__main__":
    main()

File: rust/2019/src/bin/day_02.rs
use rug::Integer;
use std::fs;

use aoc_2019::CPU;

fn run(program: &[Integer], noun: u32, verb: u32) -> Integer {
    let mut cpu = CPU::new(program);
    cpu.mem.insert(1.into(), noun.into());
    cpu.mem.insert(2.into(), verb.into());
    cpu.run();
    cpu.mem.remove(&Integer::new()).unwrap()
}

fn main() {
    let text = fs::read_to_string("../input/2019/input_02.txt").unwrap();
    let program: Vec<Integer> = text
        .trim_end()
        .split(',')
        .map(|s| s.parse().unwrap())
        .collect();
    println!("Part 1: {}", run(&program, 12, 2));
    for noun in 0..=99 {
        for verb in 0..=99 {
            if run(&program, noun, verb) == 19690720 {
                println!("Part 2: {}", 100 * noun + verb);
                return;
            }
        }
    }
}
